includes/payments/class-wc-payments-controller.php

```diff
@@ -0,0 +1,98 @@
+<?php
+/**
+ * REST API Payments Controller
+ *
+ * Handles requests to the /payments endpoint.
+ *
+ * @author   Automattic
+ * @category API
+ * @package  WooCommerce/API
+ */
+
+use Automattic\WooCommerce\Admin\Notes\Note;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * REST API Settings controller class.
+ *
+ * @package WooCommerce/API
+ */
+class WC_Payments_Controller extends WC_REST_Controller {
+
+	/**
+	 * Endpoint namespace.
+	 *
+	 * @var string
+	 */
+	protected $namespace = 'wc-calypso-bridge/v1';
+
+	/**
+	 * Route base.
+	 *
+	 * @var string
+	 */
+	protected $rest_base = 'payments';
+
+	/**
+	 * Register routes.
+	 */
+	public function register_routes() {
+		register_rest_route( $this->namespace, '/' . $this->rest_base . '/activate-promo', array(
+			array(
+				'methods'             => 'POST',
+				'callback'            => array( $this, 'activate_promo_note' ),
+				'permission_callback' => array( $this, 'check_permission' ),
+			),
+			'schema' => array( $this, 'get_public_item_schema' ),
+		) );
+	}
+
+	/**
+	 * Verify access.
+	 *
+	 * Override this method if custom permissions required.
+	 */
+	public function check_permission() {
+		return current_user_can( 'manage_woocommerce' );
+	}
+
+	/**
+	 * Set action to promo note to give the user discount eligibility.
+	 *
+	 */
+	public function activate_promo_note() {
+		$promo_name = 'wcpay-promo-2021-6-incentive-2';
+		$data_store = WC_Data_Store::load( 'admin-note' );
+		$add_where_clause = function( $where_clause ) use ( $promo_name ) {
+			return $where_clause . " AND name = '$promo_name'";
+		};
+
+		add_filter( 'woocommerce_note_where_clauses', $add_where_clause );
+		$notes = $data_store->get_notes();
+		remove_filter( 'woocommerce_note_where_clauses', $add_where_clause );
```

Comment:
I _think_ `remove_filter` is unable to remove anonymous functions like `$add_where_clause `. A bit more info [here](https://stackoverflow.com/questions/31240432/anonymous-functions-in-wordpress-hooks).

If that's the case, I think the where clause would continue to run on other notes 😬 

Would it be better to use a class method for the filter?

Reply:
I think that SO answer is talking about the case where `remove_filter` is being called in a different scope to `add_filter`, where the anonymous function  instance isn't available. Since this is calling `remove_filter` with the same instance, IMO this should work. I am a rookie however so don't trust me.

Reply:
Thank you @adrianduffell 👍 This was very interesting! I wrote the following test locally to confirm that it actually removes the filter. @becdetat I think you're right that it can be removed as long as the function can be referenced back. 

```

public function test_filter_gets_removed() {
	global $wp_filter;
	$promo_name = 'wcpay-promo-2021-6-incentive-2';
	$data_store = WC_Data_Store::load( 'admin-note' );
	$add_where_clause = function( $where_clause ) use ( $promo_name ) {
		return $where_clause . " AND name = '$promo_name'";
	};

	add_filter( 'woocommerce_note_where_clauses', $add_where_clause );
	$data_store->get_notes();
	$this->assertArrayHasKey('woocommerce_note_where_clauses', $wp_filter);

	$wheres_with_filter = $data_store->get_notes_where_clauses();
	$this->assertEquals(' AND is_deleted = 0 AND name = \'wcpay-promo-2021-6-incentive-2\'', $wheres_with_filter);

	remove_filter( 'woocommerce_note_where_clauses', $add_where_clause );

	$wheres_without_filter = $data_store->get_notes_where_clauses();
	$this->assertEquals(' AND is_deleted = 0', $wheres_without_filter);

	$this->assertArrayNotHasKey('woocommerce_note_where_clauses', $wp_filter);
}
```

Reply:
Awesome stuff, thanks for investigating it!